Doc/reference/datamodel.rst

```diff
@@ -2590,6 +2591,19 @@ through the object's keys; for sequences, it should iterate through the values.
    the container.
 
 
+.. method:: object.keys(self)
+
+   This method can be used by Python in arbitrary circumstances to treat an
```

Comment:
«Arbitrary» makes it sound like this is not well defined

Reply:
That's the goal : to make no promises as to which part of the implementation use `keys` rather than, for example, `__iter__`. That formulation could disappear if we consider what I wrote as an implementation detail to be documented, and if we also make the pledge not to rely on `keys` for **anything** other than the unary `**` and the update method. I think it's better this way.

Reply:
But it sends a mixed message to have the datamodel doc sound uncertain about this!

Reply:
Are there any places in Python where the presence of `keys` is *not* used to determine a mapping?

Reply:
@ethanfurman there aren't many places where it needs to be determined whether something is a mapping or not, so I guess no, not really. Though `issubclass`/`isinstance` with collections.abc.MutableMapping doesn't check that and only checks inheritance (and things registered to it using the class's register method), as opposed to other ABCs in the same module - but that's the stdlib, not the CPython implem itself.

Reply:
Maybe we could say that Python calls that method when it needs to tell a mapping apart from a sequence, and that it happens at least with the unary `**` and in dict.update ? Without implementation detail entry at all. And assuming no other instance where it doesn't check `keys` is found.